packages/@core.keystone/cache.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Keystone-level cache.
+ *
+ * This package adds basic caching capabilities to keystone.
+ *
+ * Requirements:
+ * 1. Every request should have x-request-id. If it doesn't have it -- this request will not use cache
+ *
+ * Installation:
+ * 1. Have a custom keystone server
+ * 2. Somewhere in index.js call initCache(keystone, cache)
+ * 3. Add a middleware KeystoneLevelCache(cache) before keystone AdminApp or GraphQL app
+ *
+ * How it works:
+ *                           -> [GQL Request] -> [DB]
+ * [HTTP Request] -> [Cache] -> [GQL Request] -> [DB]
+ *                           -> [GQL Request] -> [DB]
+ *
+ * On every incoming request module simply checks x-request-id header and operation.
+ * If request is `read` - return cache item (assuming cache has this item)
+ * If request has other type - remove cache item by request id
+ *
+ * Cache structure:
+ * cache: {
+ *     <x-request-id>: {
+ *      	<request-args>: []
+ *      	<request-args>: []
+ *      	<request-args>: []
+ *     }
+ * }
+ *
+ */
+
+const express = require('express')
+const { get } = require('lodash')
+
+class KeystoneCacheMiddleware {
+
+    constructor (cache) {
+        if (!(typeof cache === 'object') || cache === null) {
+            throw new Error('Cache should be dict!')
+        }
+        this.cache = cache
+    }
+
+    prepareMiddleware ({ keystone, dev, distDir }) {
+
+        const app = express()
+        app.use((req, res, next) => {
+            const requestId = get(req, ['headers', 'x-request-id'])
+            res.on('close', () => {
+                if (requestId) { delete this.cache[requestId] }
+            })
+            next()
+        })
+        return app
+    }
+}
+
+const generateRequestKey = (gqlName, args) => (
+    `${gqlName}-${JSON.stringify(args)}`
+)
+
+const getRequestIdFromContext = (context) => {
+    return get(context, ['req', 'headers', 'x-request-id'], null)
+}
+
+const patchMutation = (mutationContext, mutation, isUpdateMutation, cache) => {
+    return async (data, context, mutationState) => {
+        let requestId
+        if (isUpdateMutation) {
+            requestId = getRequestIdFromContext(mutationState)
+        } else {
+            requestId = getRequestIdFromContext(context)
+        }
```

Comment:
why do we need this condition?

Reply:
For some reason, keystone's internal `list.createMutation` and `list.updateMutation` have different signature, thus the condition

Reply:
Probably it’s better to split this logic too?